keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+  2. Set `spec.preferDualStack: true` api-server will default `spec.ipFamilies` according to how they are configured on the cluster.
+  2. Set `spec.preferDualStack: true` and set `spec.ipFamilies` to max of one ipfamily. This ipfamily will be treated as the primary ip family for the service. api-server will default `spec.ipFamilies: [ 'family provided by user', 'other family configured on cluster, if any']`
+  3. Set `spec.preferDualStack: true` and set `spec.ClusterIPs` to max of one ip. This ip will be treated as the primary ip for the resultant service (if available for allocation). api-server will default `spec.ipFamilies` and set `spec.ClusterIPs: ['ip provided by user', 'allocated ip from cluster secondary ip family']`
+
 
-Dual-stack services (having both an IPv4 and IPv6 cluster IP) are currently out of scope of this proposal. Should there be use-cases where dual-stack Services are needed then we can revisit.
+The below are sample services that demonstrate the above rules
+
+```yaml
+# dual stack service (required)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user
+  ipFamilies: # set by user
+    - IPv4
+    - IPv6
+  clusterIP: 1.2.3.4  # allocated
+  clusterIPs: # allocated
+    - 1.2.3.4
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (required)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # defaulted
+  ipFamilies: # defaulted
+    - IPv6
+    - IPv4
+  clusterIP: 2001::1  # set by user
+  clusterIPs: # set by user
+    - 2001::1
+    - 1.2.3.4
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (optional)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user 
+  ipFamilies:
+    - IPv6 # defaulted, cluster does not support dual stack services, only one ipv6 cidr is provided for services
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1 # note single entry
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (optional)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user 
+  ipFamilies:
+    - IPv6 # set by user
+    - IPv4 # defaulted (cluster supports dual stack)
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1
+    - 1.2.3.4
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+```
+
+##### Notes on NodePort Allocation
+
+NodePort reservation will be constant across both families. That mean a reservation of `NodePort: 12345` will happen across both families irrespective of the ip family of the service.
```

Comment:
detail: NodePorts will only forward traffic for families that are actually assigned to that service